src/Language/PureScript/Pretty/JS.hs

```diff
@@ -58,9 +57,9 @@ literals = mkPattern' match'
     , return $ emit "}"
     ]
     where
-    objectPropertyToString :: (Emit gen) => Text -> gen
-    objectPropertyToString s | identNeedsEscaping s = string s
-                             | otherwise = emit s
+    objectPropertyToString :: (Emit gen) => PSString -> gen
+    objectPropertyToString s | identNeedsEscaping $ T.pack $ codePoints s = string s
```

Comment:
While this implementation will do the right thing as far as I can tell, it makes me a bit uneasy since `T.pack . codePoints` can introduce replacement characters. If I understand correctly, I think the following is a bit closer to how I might describe what should happen here:

```haskell
-- | Attempt to decode a PSString as UTF-16 encoded text. This will fail if e.g. the string contains lone surrogates.
codePoints :: PSString -> Maybe Text

objectPropertyToString s =
  case codePoints s of
    Just s' | not (identNeedsEscaping s') ->
      emit s'
    _ ->
      string s
```

(perhaps if we were to do it this way, `decode` or `decodeText` might be a better name for `codePoints`).

That is, the only case where we can safely emit an object property without using a string literal is when a) the string does represent valid UTF-16 encoded text and b) that text is a valid JS identifier (i.e. doesn't contain primes or anything like that). In all other cases, we need to emit it as a string literal.

I can't remember if I mentioned this already but exporting `codePoints` without using `Maybe` in the return type makes me uneasy because it's not always possible to consider a `PSString` as a sequence of code points (as you've already pointed out elsewhere). Is it feasible to use `Maybe`?